torchdata/dataloader2/communication/protocol.py

```diff
@@ -225,6 +225,16 @@ def get_response_reset_iterator(self, block=False):
         if not isinstance(response, communication.messages.ResetIteratorResponse):
             raise Exception("Invalid response received")
 
+    def get_response_reset_epoch(self, block=False):
```

Comment:
Not sure why we didn't have this response before

Reply:
Non-blocking but - hmmmm.... did this cause any bug or unhandled messages? Do you happen to know why?
I know you were looking into unusually messages and responses

Reply:
When new iteration starts, `reset` will be called for `_IterateQueueDataPipes `. And, an extra `get_respondse_next` is invoked to drop / response. So, all requests are served.
https://github.com/pytorch/data/blob/0a0ae5d35be439c786c9ab798ea1596238d8640a/torchdata/dataloader2/reading_service.py#L152-L154

I can give a try to remove this part.

Reply:
It seems removing is fine

Reply:
noob question: any docs / entry pointer to understand how these `dataloader2/communication` design / work? ~

Reply:
Unfortunately no doc. I can talk about the components based on my understanding:
- [ProtocolClient](https://github.com/pytorch/data/blob/main/torchdata/dataloader2/communication/protocol.py#L196) remains in the main process to pass `Request` via the `request_queue` to the corresponding worker process
- [ProtocolServer](https://github.com/pytorch/data/blob/56ed97e48da9934ce23453fc7b04418b64b02e8a/torchdata/dataloader2/communication/protocol.py#L168) is created in the worker process  that takes request then send `Response` back to main process  via `reqponse_queue`
- [DataPipeBehindQueues](https://github.com/pytorch/data/blob/main/torchdata/dataloader2/communication/iter.py#LL109C5-L109C25) is the worker loop that holds a ProtocolServer to maniputlate `DataPipe` based on the `Reqeust`
- [QueueWrapper](https://github.com/pytorch/data/blob/main/torchdata/dataloader2/communication/iter.py#L170) is the `DataPipe` that holds a `ProtocolClient` instance to issue `Request` and yield data from `response_queue` to the subsequent `DataPipe` graph.

We can talk about more detail offline if you want

Reply:
Thanks! Something like this? ~

![319563983_491898436260282_330096271851500173_n](https://user-images.githubusercontent.com/799346/208835936-af80c20f-beff-4a24-ab82-414b5c9d88d8.jpg)

Reply:
Correct.

Reply:
Let me try to draw it with mermaid

```mermaid
graph TD;
    Worker1_1-->ProtocolServer_1_1-->ProtocolClient_1
    Worker1_2-->ProtocolServer_1_2-->ProtocolClient_1
    Worker1_3-->ProtocolServer_1_3-->ProtocolClient_1
    ProtocolClient_1-->GPU1
    ProtocolClient_2-->GPU2
    ProtocolClient_3-->GPU3
```